### ISSUE TITLE
ArgumentParser's help does not show the list of valid extensions for parameters

### ISSUE THREAD

**Issue by esiragusa** [seqan/seqan#838]
Check the output of any seqan app with --help.

**Comment by esiragusa** [comment]
@h-2 we forgot to address this issue in #1162!

**Comment by h-2** [comment]
@esiragusa From my POV the app developers just need to set it right. This is how lambda does it:
```c++
setValidValues(parser, "database", toCString(concat(getFileExtensions(SeqFileIn()), ' ')));
```

**Comment by h-2** [comment]
Maybe someone should look through the apps and replace lines like
https://github.com/seqan/seqan/blob/master/apps/mason2/mason_genome.cpp#L118
with the above line.

@xp3i4 can you do this?

**Comment by xp3i4** [comment]
@h-2 So we need to replace extensions using string directly with getFileExtensions() function for all apps?

**Comment by h-2** [comment]
@xp3i4 yes, wherever there is hardcoded "fasta fastq" instead use getFileExtensions()!

**Comment by esiragusa** [comment]
@h-2 But I already did that (see [here](https://github.com/seqan/seqan/blob/master/apps/yara/mapper.cpp#L120)). The help screen doesn't show which extensions are allowed.

@xp3i4 please change all other apps using ```setValidValues(parser, "database", SeqFileIn::getFileExtensions());``` instead of using the version proposed above by @h-2.

**Comment by h-2** [comment]
@esiragusa why should @xp3i4 use your version if it doesn't work? :trollface: 
My version works:
```
    -q, --query IN
          Query sequences. Valid filetypes are: .fq, .fq.bgzf, .fq.gz, .fq.bz2, .fastq, .fastq.bgzf, .fastq.gz,
          .fastq.bz2, .fa, .fa.bgzf, .fa.gz, .fa.bz2, .fasta, .fasta.bgzf, .fasta.gz, .fasta.bz2, .faa, .faa.bgzf,
          .faa.gz, .faa.bz2, .ffn, .ffn.bgzf, .ffn.gz, .ffn.bz2, .fna, .fna.bgzf, .fna.gz, .fna.bz2, .frn, .frn.bgzf,
          .frn.gz, .frn.bz2, .embl, .embl.bgzf, .embl.gz, .embl.bz2, .gbk, .gbk.bgzf, .gbk.gz, .gbk.bz2, .raw,
          .raw.bgzf, .raw.gz, and .raw.bz2.
```

**Comment by esiragusa** [comment]
@h-2 I think you got confused. This issue concerns ArgParse parameters (1, 2, etc.) not ArgParse arguments (--database). I guess that your version won't work either with parameters. The version I proposed above works with arguments and it is more concise.

**Comment by esiragusa** [comment]
Look here:

```
$ yara_mapper --help
yara_mapper - Yara Mapper
=========================

SYNOPSIS
    yara_mapper [OPTIONS] <REFERENCE INDEX PREFIX> <SE-READS FILE>
    yara_mapper [OPTIONS] <REFERENCE INDEX PREFIX> <PE-READS FILE 1> <PE-READS FILE 2>

DESCRIPTION
...
  Output Options:
    -o, --output-file FILE
          Specify an output file. Default: write the file to standard output. Valid filetypes are: .bam, .sam,
          .sam.bgzf, .sam.gz, and .sam.bz2.
...
```

The help screen of yara_mapper doesn't show valid extensions for index reads parameters, while it shows extensions for other arguments.

**Comment by h-2** [comment]
> @h-2 I think you got confused.

I did, thanks for clarifying. I never used "parameters", only arguments because they are agnostic of order.

**Comment by h-2** [comment]
@xp3i4 what is the state of this?

**Comment by smehringer** [comment]
@h-2 @esiragusa 

So as I see it there are two possible ways to fix the issue for parameters (I will change the valid file extension for arguments as you proposed) and I wanted to ask which one you prefer before working on it:

1. Don't use parameters but arguments. :trollface: (I could change all apps accordingly, introducing e.g. -i or --index-prefix for each parameter)

2. Add a global section to the argument parsers help page like *"Parameter Restrictions"* listing the valid values for each parameter if given.

**Comment by h-2** [comment]
I would prefer 2.

**Comment by smehringer** [comment]
Since @esiragusa didn't answer yet I worked on solution 2:

I also noticed that neither an arguments numerical range, default values or helpText is displayed (if specified at all). So I included those too. 

Additionally the help page did not inform the user whether the argument or option is a List. I changed the parser according to display ```List of [type]'s``` or ```List of 2 [type]'s``` instead of ```[type]``` whenever isList or numberOfValues is specified 

@h-2 @esiragusa What do you think:
![bildschirmfoto - 28 10 2016 - 13 35 41](https://cloud.githubusercontent.com/assets/7955878/19805233/751f4aba-9d13-11e6-891b-2ce00cbcf686.png)

**Comment by smehringer** [comment]
Note that I used the words Options and Arguments consistently with the naming in our documentation (not parameters and arguments as in your comments above).

**Comment by esiragusa** [comment]
@smehringer @h-2 That's perfect! Thank you 😄

**Comment by esiragusa** [comment]
Shouldn't PREFIX be something like FILENAME_PREFIX?

**Comment by smehringer** [comment]
Hm yes, or maybe INDEX_PREFIX. I'll have a look if I can change that (it automatically detects the argument input types somewhere)

**Comment by xp3i4** [comment]
Thank you @smehringer , I was stuck in the issue for a long time.

**Comment by smehringer** [comment]
I'm happy to help :smile:

**Comment by smehringer** [comment]
Fixed by #1901
